src/txmempool.cpp

```diff
@@ -108,18 +122,23 @@ void CTxMemPool::UpdateForDescendants(txiter update_it, cacheMap& cache, const s
     int64_t modify_size = 0;
     CAmount modify_fee = 0;
     int64_t modify_count = 0;
+    // All entries in the update_cache have
+    // already been checked against excluded list
     for (txiter child_it : update_cache) {
         const CTxMemPoolEntry& child = *child_it;
-        if (!exclude.count(child.GetTx().GetHash())) {
-            modify_size += child.GetTxSize();
-            modify_fee += child.GetModifiedFee();
-            modify_count++;
-            mapTx.modify(child_it, update_ancestor_state(update_it->GetTxSize(), update_it->GetModifiedFee(), 1, update_it->GetSigOpCost()));
-        }
+        modify_size += child.GetTxSize();
+        modify_fee += child.GetModifiedFee();
+        modify_count++;
+        mapTx.modify(child_it, update_ancestor_state(update_it->GetTxSize(), update_it->GetModifiedFee(), 1, update_it->GetSigOpCost()));
     }
     mapTx.modify(update_it, update_descendant_state(modify_size, modify_fee, modify_count));
     // share the cache (if there is one)
-    if (!update_cache.empty()) cache.emplace(update_it, std::move(update_cache));
+    if (!update_cache.empty()) {
+        // Ask to shrink to fit if we're using space poorly
+        if (update_cache.size() > 100 && update_cache.capacity() > 2 * update_cache.size())
```

Comment:
I overlooked this issue in #18120 -- is this the regression you're referring to in the commit message?

Reply:
Nope this is *as a result* of the fix, we sometimes now have a cache that is too big because we remove things from the cache if they are in exclude. We could not do this -- and then it would still be possible for the cache entries to have an excess of O(direct children in exlcude), rather than O(exclude) without this patch, and an excess of 100 or not more than 2*size with this patch. It also only does anything on systems that implement shrink_to_fit fwiw.


I'll tag you on the relevant line....